### ISSUE TITLE
More helpful error messages when plan_stubs are run without enough context

### ISSUE THREAD

**Issue by danielballan** [bluesky/bluesky#1240]
Diverting discussion from #1238. Can we make this error message more specific or prescriptive?

```python
In [3]: RE(trigger_and_read())
<snipped>
IllegalMessageSequence: Cannot bundle readings without an open run. That is, 'create' must be preceded by 'open_run'.
```

**Comment by ZLLentz** [comment]
If there was a way for a plan to request the current state of the run engine with a message, it would be possible to write custom error messages in the stubs that were more specific to the situation. This would slow down these plans overall by adding extra "junk" messages that often just double-check if a run is open (to see if we need to raise our custom error message), but I think it would solve cases like these in a maintainable way.

```
def my_stub():
    re_state = yield from run_engine_state()
    if re_state.is_open == False:
        raise MyStubException("my_stub is meant to be used in...")
```

**Comment by danielballan** [comment]
Oh, great suggestion! Riffing on this, we can get more specific error messages without even adding a new message:

```python
import bluesky.plan_stubs
import bluesky
from bluesky import RunEngine
from ophyd.sim import det
RE = RunEngine()

def improved_trigger_and_read(detectors):
    try:
        yield from bluesky.plan_stubs.trigger_and_read(detectors)
    except bluesky.IllegalMessageSequence:
        raise RuntimeError("The trigger_and_read plan must be used inside an open run.")

RE(improved_trigger_and_read([det]))
```

Tested interactively, I get:

```python
In [10]: RE(improved_trigger_and_read([det]))
<snipped>
RuntimeError: The trigger_and_read plan must be used inside an open run.
```

That does seem much better!

To refine this, I would make the `try/except` part of `trigger_and_read` _itself_ rather than wrapping it as I did above, and I would consider making the exceptions raised by RE more specific (subclasses of `IllegalMessageSequence` for each situation?) to ensure that we can catch _exactly_ the case that we mean to catch.

**Comment by danielballan** [comment]
(What do you think, @ZLLentz? Can you see any downsides compared to your original suggestion of adding a new message type?)

**Comment by ZLLentz** [comment]
I think my original suggestion has potential for other utilities, but more specific exceptions from the RE is a perfectly good and very useful solution. :+1:

The only downside I see is that we can't know anything in particular about the context except for the type of the exception and some exception-specific data, whereas with the other route we can add anything and everything the run engine knows about. Some upsides to your solution is that it doesn't add these junk messages to normal operations and doesn't require us to write and maintain a new message type.

**Comment by danielballan** [comment]
I like the pattern of stashing useful context onto exception instances, something which I think is generally underused. See this example from event-model:

```python
                    err_with_key = UnresolvableForeignKeyError(
                        f"Event with uid {doc['uid']} refers to unknown Datum "
                        f"datum_id {datum_id}")
                    err_with_key.key = datum_id
```

Would it help to stash the RE state (and anything else...) onto the `IllegalMessageSequence` or subclass thereof? It's still not quite as flexible as your proposal, but it's interesting to explore how close we can get to your use case without going all the way to a new proactive state-check Msg.

**Comment by klauer** [comment]
Would there be any difference here if you could query the run engine to see if a message was valid in the current run-engine state proactively?

```python
valid = yield Msg('is_this_a_valid_message_in_the_current_state', 'create')
```

I feel like the answer is "not really" (why not just send the actual message?), but thought I'd bring it up regardless.

**Comment by danielballan** [comment]
Yeah, I think the answer is, "Send the actual message, and if you like handle the error and send a different message."

**Comment by ZLLentz** [comment]
It makes sense to add this kind of context information into the `IllegalMessageSequence` instance as a convenience, though the full scope of what is "useful" context information is something that will change over time. If this context information ends up being very useful in a few cases, I would ask that we consider adding the message too so that I don't have to intentionally cause an error when I want the information. However I don't currently have a use-case for this so your solution is best.

**Comment by danielballan** [comment]
I think we're on exactly the same page with this then. Your suggestion sounds like a helpful thing to add if we encounter a use case to justify the minor but nonzero impact on complexity/maintenance/performance. My suggestion is sufficient to address the problem that prompted this GitHub issue.

**Comment by ZLLentz** [comment]
Yes, I agree. In fact the context information isn't even important in this case, just the splitting of the error types so we know that an `IllegalBundlingSequence` error occurred. I would be on board for only doing the parts that we have current use-cases for.

**Comment by tacaswell** [comment]
One of the mid-term goals is to make the RE more modular (both from a cleanly ness point of view, but also to support mulitple open runs at the same time) so "the state" of the RE is going going to get more complex.  Following the python, I think a "ask forgiveness" path is better as it is easier to make sure the RE is consistent about how it says "No" rather than on plans knowing how to figure out if the RE _would_ say "No".

**Comment by awalter-bnl** [comment]
@danielballan asked me to take a look at opening a PR for this, I just wanted to make sure I understood the full scope of said PR before starting. My proposal is:

1. Add new child exception classes: IllegalRunSequence, IllegalBundlingSequence, IllegalMonitorSequence (based on their use in 'if' loops in the RunEngine that check against ``self._run_is_open``, ``self._bundling``, ``self._monitor_params``) Are there others people would like to see, perhaps IllegalRunNotOpenSequence/IllegalRunOpenSequence ?
1. Update the RunEngine to use the new exception classes above.
1. Update any 'plan_stubs' that use more than one base message that has any of the exceptions above in it with a ``try``, ``except`` block that re-issues the message indicating that it is the cause of the exception (trigger_and_read, one_1D_step, move_per_step, one_nd_step, repeat, ...).

I am less clear about _what_ items we want to 'stash' onto the exceptions as attributes, Any suggestions here can be implemented.

@danielballan, @tacaswell, @ZLLentz, @klauer  is this what you guys had in mind?

**Comment by danielballan** [comment]
Exactly. We might want to iterate on the names in (1), but that be done easily enough with Find/Replace during PR review.

Things I can think to stash:
* `currently_open_runs` (list of run start UIDs, will just have one element for now but might have multiple in future)
* `run_engine_state`

**Comment by ZLLentz** [comment]
Also, there's no reason to assume this PR needs to stash the full set of things anyone ever might need forever so long as it's simple to extend it as needed. I think you should start with stashing information that is immediately useful for the plan stubs error messages.
